### PR TITLE
[Community] migrate to swell backend

### PR DESCRIPTION
**Pull request by GuillaumeDSM**

### CODE REVIEW

File: octobot/community/authentication.py [hunk 1]
```diff
@@ -185,11 +216,23 @@ def _update_supports(self, resp_status, json_data):
             self.logger.error(f"Error when fetching community support, "
                               f"error code: {resp_status}")
 
+    def _supports_mock(self):
+        return {
+            "data": {
+                "attributes": {
+                    "support_role": "OctoBot tester"
```
Comment (Herklos) [inline octobot/community/authentication.py:223]: 😄

**Review by Herklos — APPROVED** [review]
Nice!

File: octobot/community/feeds/community_mqtt_feed.py [hunk 2]
```diff
@@ -77,18 +90,100 @@ async def register_feed_callback(self, channel_type, callback, identifier=None):
...
+        """
+        create_resp = await self.authenticator.async_graphql_query(create_query)
+        if create_resp.status == 200:
+            self._device_id = (await create_resp.json())["data"]["createDevice"]["id"]
+        else:
+            raise RuntimeError(
+                f"Error when creating mqtt device, code: {create_resp.status}, text: {await create_resp.text()}"
+            )
+
+    async def _fetch_device_uuid(self):
```
Comment (Herklos) [inline octobot/community/feeds/community_mqtt_feed.py:110]: 👍

File: octobot/community/authentication.py [hunk 3]
```diff
@@ -94,29 +117,105 @@ async def send(self, message, channel_type, identifier=None):
...
+                if resp.status == 401:
+                    # access token expired
+                    raise authentication.AuthenticationRequired
+                json_resp = await resp.json()
+                if errors := json_resp.get("errors"):
+                    raise RuntimeError(f"Error when running graphql query: {errors[0].get('message', errors)}")
+                return resp
+        except authentication.AuthenticationRequired:
+            if allow_retry_on_expired_token:
+                return await self.async_graphql_query(query, variables=variables, operation_name=operation_name,
```
Comment (Herklos) [inline octobot/community/authentication.py:147]: 👍

File: octobot/community/authentication.py [hunk 4]
```diff
@@ -94,29 +117,105 @@ async def send(self, message, channel_type, identifier=None):
...
+            else:
+                raise
+
+    @contextlib.asynccontextmanager
+    async def _authenticated_qgl_session(self):
+        await self.gql_login_if_required()
+        session = self.get_aiohttp_session()
+        try:
+            session.headers[self.GQL_AUTHORIZATION_HEADER] = f"Bearer {self._gql_access_token}"
+            yield session
```
Comment (Herklos) [inline octobot/community/authentication.py:158]: 👍

File: octobot/community/authentication.py [hunk 5]
```diff
@@ -94,29 +117,105 @@ async def send(self, message, channel_type, identifier=None):
...
     def can_authenticate(self):
         return "todo" not in self.authentication_url
 
-    def login(self, username, password):
+    async def login(self, username, password):
         self._ensure_community_url()
         self._reset_tokens()
         params = {
-            "username": username,
+            "email": username,
```
Comment (Herklos) [inline octobot/community/authentication.py:193]: Shouldn't we rename `username` to `email` ?

File: octobot/community/feeds/community_mqtt_feed.py [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+    DEVICE_CREATE_TIMEOUT = 5 * commons_constants.MINUTE_TO_SECONDS
+    DEVICE_CREATION_REFRESH_DELAY = 2
+
+    # Quality of Service level determines the reliability of the data flow between a client and a message broker.
+    # The message may be sent in three ways:
+    # QoS 0: the message will be received at most once (also known as “fire and forget”).
+    # QoS 1: the message will be received at least once.
+    # QoS 2: the message will be received exactly once.
+    # from https://www.scaleway.com/en/docs/iot/iot-hub/concepts/#quality-of-service-levels-(qos)
+    default_QOS = 1
```
Comment (Herklos) [inline octobot/community/feeds/community_mqtt_feed.py:48]: 👍

File: octobot/community/feeds/community_mqtt_feed.py [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
+            raise RuntimeError(f"{error_message}, {e}")
+        if create_resp.status == 200:
+            json_resp = await create_resp.json()
+            return json_resp["data"]["insertOneDevice"]["_id"]
+        else:
+            raise RuntimeError(f"{error_message}, code: {create_resp.status}, text: {await create_resp.text()}")
+
+    async def _fetch_device_uuid(self, device_id):
+        select_query = """
+        query SelectDeviceUUID($_id: ObjectId, $user_id: ObjectId) {
```
Comment (Herklos) [inline octobot/community/feeds/community_mqtt_feed.py:112]: 👍

Comment (Herklos) [inline octobot/community/feeds/community_mqtt_feed.py:112]: Shouldn't we store queries and mutations in a separate module?

File: octobot/constants.py [hunk 8]
```diff
@@ -44,12 +45,21 @@
...
 OCTOBOT_COMMUNITY_FETCH_FEED_IDENTIFIER_URL = f"{OCTOBOT_COMMUNITY_URL}api/v2/storefront/feeds/id"
 COMMUNITY_FEED_CURRENT_MINIMUM_VERSION = "1.0.0"
+COMMUNITY_FEED_DEFAULT_TYPE = octobot.enums.CommunityFeedType.MQTTFeed
+
+COMMUNITY_BACKEND_API_URL = "https://octo-bot.swell.store/api"
+COMMUNITY_GQL_BACKEND_API_URL = "https://eu-west-1.aws.realm.mongodb.com/api/client/v2.0/app/community-backend-production-wgdgz/graphql"
+COMMUNITY_GQL_AUTH_URL = "https://realm.mongodb.com/api/client/v2.0/app/community-backend-production-wgdgz/auth/providers/api-key/login"
+COMMUNITY_GQL_REFRESH_AUTH_URL = "https://realm.mongodb.com/api/client/v2.0/auth/session"
+COMMUNITY_BACKEND_AUTH_URL = f"{COMMUNITY_BACKEND_API_URL}/account/login"
+COMMUNITY_TOKEN = os.getenv("COMMUNITY_TOKEN", "pk_mvKn5XznfVVSyp3XZlgodBKWlRhazeSu")
```
Comment (Herklos) [inline octobot/constants.py:61]: Did you commit your own token or it's a public one?

File: octobot/community/feeds/community_mqtt_feed.py (hunk 7 above)
Comment (GuillaumeDSM) [reply octobot/community/feeds/community_mqtt_feed.py:112]: right

File: octobot/constants.py (hunk 8 above)
Comment (GuillaumeDSM) [reply octobot/constants.py:61]: it's public :)

**Pull request merged**